narwhal/primary/src/primary.rs

```diff
@@ -591,6 +585,61 @@ impl PrimaryToPrimary for PrimaryReceiverHandler {
         .map(|_| anemo::Response::new(()))
         .map_err(|e| anemo::rpc::Status::internal(e.to_string()))
     }
+
+    async fn fetch_certificates(
+        &self,
+        request: anemo::Request<FetchCertificatesRequest>,
+    ) -> Result<anemo::Response<FetchCertificatesResponse>, anemo::rpc::Status> {
+        let request = request.into_body();
+        let mut response = FetchCertificatesResponse {
+            certificates: Vec::new(),
+        };
+        if request.progression.is_empty() || request.max_items == 0 {
+            return Ok(anemo::Response::new(response));
+        }
+
+        // Use a min-queue for (round, authority) to keep track of the next certificate to fetch.
+        // If the requestor is providing its current round on all authorities, it should be able to
+        // process certificates returned in this order without any missing parents.
+        //
+        // Compared to fetching certificates iteratatively round by round, using a heap is simpler,
+        // and avoids the pathological case of iterating through many missing rounds of a downed authority.
+        let mut fetch_queue = BinaryHeap::new();
+        for (authority, round) in request.progression.into_iter() {
+            let next_round = self
+                .certificate_store
+                .next_round_number(&authority, round)
+                .map_err(|e| anemo::rpc::Status::from_error(Box::new(e)))?;
+            if let Some(r) = next_round {
+                fetch_queue.push(Reverse((r, authority)));
+            }
+        }
+
+        while let Some(Reverse((round, authority))) = fetch_queue.pop() {
+            match self
+                .certificate_store
+                .read_by_index(authority.clone(), round)
+                .map_err(|e| anemo::rpc::Status::from_error(Box::new(e)))?
+            {
+                Some(cert) => {
+                    response.certificates.push(cert);
+                    let next_round = self
+                        .certificate_store
+                        .next_round_number(&authority, round)
+                        .map_err(|e| anemo::rpc::Status::from_error(Box::new(e)))?;
+                    if let Some(r) = next_round {
+                        fetch_queue.push(Reverse((r, authority)));
+                    }
+                }
+                None => continue,
+            };
+            if response.certificates.len() == request.max_items {
```

Comment:
same suggestion, >= here?

Reply:
The flip side of using `>=` is that readers would need to think to see if there is indeed a case that `response.certificates.len() > request.max_items`. I added an assert to make sure `max_items` is respected.